src/main/java/jp/co/soramitsu/sora/crypto/service/impl/JSONVerifierImpl.java

```diff
@@ -0,0 +1,54 @@
+package jp.co.soramitsu.sora.crypto.service.impl;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.io.IOException;
+import java.security.Signature;
+import java.security.SignatureException;
+import jp.co.soramitsu.sora.common.Util;
+import jp.co.soramitsu.sora.crypto.common.Consts;
+import jp.co.soramitsu.sora.crypto.proof.Options;
+import jp.co.soramitsu.sora.crypto.proof.Proof;
+import jp.co.soramitsu.sora.crypto.service.JSONCanonizer;
+import jp.co.soramitsu.sora.crypto.service.JSONVerifier;
+import lombok.AllArgsConstructor;
+
+
+@AllArgsConstructor
+public class JSONVerifierImpl implements JSONVerifier {
+
+  private final JSONCanonizer canonizer;
+  private final ObjectMapper mapper;
+
+  @Override
+  public boolean verify(Object object, Signature signature)
+      throws IOException, SignatureException {
+    ObjectNode root = mapper.valueToTree(object);
+
+    if (!root.hasNonNull(Consts.PROOF_KEY)) {
+      return false;
+    }
+
+    // read proof
+    JsonNode proofNode = root.get(Consts.PROOF_KEY);
+    Proof proof = mapper.treeToValue(proofNode, Proof.class);
+    Options options = new Options(
+        proof.getType(),
+        proof.getCreated(),
+        proof.getCreator(),
+        proof.getNonce(),
+        proof.getPurpose()
+    );
+
+    Util.verifyAlgorithmType(options, signature);
+
+    // sanitize and serialize JSON
```

Comment:
I prefer to avoid comments in methods. 
For me, if I want to add some comments, then it is a signal that my code requires additional explanations. After that, I usually extract the code into a method and name it so that it would not add comments (self-documenting code) Sometimes, it leads to a restructuring of the app's architecture. 
I know you know it, I just explain my intentions to refactor the code

I cannot recommend you something specific because security is Greek to me (темный лес :) ) 
But I will try to discover it more closely on the weekend